File: algorithmic_toolbox/week2_algorithmic_warmup/5_fibonacci_number_again/fibonacci_huge.go
package main

import (
	"C"
	"fmt"
)

func getFibonacciHugeNaive(n, m int64) int64 {
	if n <= 1 {
		return n
	}

	var previous, current int64 = 0, 1

	for i := int64(0); i < n-1; i++ {
		tmpPrevious := previous
		previous = current
		current = (tmpPrevious + current) % m
	}

	return current
}

func getPisanoPeriod(m int64) int64 {
	var a, b int64 = 0, 1
	var c int64 = a + b
	var period int64 = 0

	for i := int64(0); i < m*m; i++ {
		c = (a + b) % m
		a = b
		b = c

		if a == 0 && b == 1 {
			period = i + 1
			break
		}
	}

	return period
}

func getFibonacciHuge(n, m int64) int64 {
	pisanoPeriod := getPisanoPeriod(m)
	remainder := n % pisanoPeriod
	return getFibonacciHugeNaive(remainder, m)
}

func main() {
	var n, m int64
	fmt.Scanf("%d %d", &n, &m)

	fmt.Println(getFibonacciHuge(n, m))
	fmt.Println(getFibonacciHugeNaive(n, m))
}

File: algorithmic_toolbox/week2_algorithmic_warmup/5_fibonacci_number_again/fibonacci_huge.py
import sys


def get_fibonacci_huge_naive(n: int, m: int) -> int:
    if n <= 1:
        return n

    previous = 0
    current = 1

    for _ in range(n - 1):
        previous, current = current, previous + current

    return current % m


def get_pisano_period(m: int) -> int:
    a, b = 0, 1
    c = a + b

    for i in range(m * m):
        c = (a + b) % m
        a = b
        b = c

        if a == 0 and b == 1:
            return i + 1

    raise Exception("could not find pisano period")


def get_fibonacci_huge(n: int, m: int) -> int:
    return get_fibonacci_huge_naive(n % get_pisano_period(m), m)


if __name__ == "__main__":
    n, m = map(int, sys.stdin.read().split())
    print(get_fibonacci_huge(n, m))
    print(get_fibonacci_huge_naive(n, m))
